include/crypto/mlkem.h

```diff
@@ -0,0 +1,166 @@
+                                            size_t seed_len);
+
+    /*
+     * ossl_mlkem768_public_from_private sets |*out_public_key| to the public key that
+     * corresponds to |private_key|. (This is faster than parsing the output of
+     * |ossl_mlkem768_generate_key| if, for some reason, you need to encapsulate to a key
+     * that was just generated.)
+     */
+    void ossl_mlkem768_public_from_private(ossl_mlkem768_public_key *out_public_key,
+                                           const ossl_mlkem768_private_key *private_key);
+
+  /* ossl_mlkem1024_CIPHERTEXT_BYTES is number of bytes in the ML-KEM-1024 ciphertext. */
+#  define OSSL_MLKEM1024_CIPHERTEXT_BYTES 1568
+
+  /* ossl_mlkem768_CIPHERTEXT_BYTES is number of bytes in the ML-KEM-768 ciphertext. */
+#  define OSSL_MLKEM768_CIPHERTEXT_BYTES 1088
+
+  /* ossl_mlkem768_SHARED_SECRET_BYTES is the number of bytes in an ML-KEM shared secret. */
+#  define OSSL_MLKEM768_SHARED_SECRET_BYTES 32
+
+    /*
+     * ossl_mlkem768_encap encrypts a random shared secret for |public_key|, writes the
+     * ciphertext to |out_ciphertext|, and writes the random shared secret to
+     * |out_shared_secret|.
+     * it is assumed out_ciphertext has been allocated ossl_mlkem768_CIPHERTEXT_BYTES bytes
+     * and out_shared_secret has been allocated MLKEM_SHARED_SECRET_BYTES bytes
+     */
+    int ossl_mlkem768_encap(uint8_t *out_ciphertext,
+                            uint8_t *out_shared_secret,
+                            const ossl_mlkem768_public_key *public_key);
+
+    /*
+     * ossl_mlkem768_decap decrypts a shared secret from |ciphertext| using |private_key|
+     * and writes it to |out_shared_secret|. If |ciphertext_len| is incorrect it
+     * returns 0, otherwise it returns 1. If |ciphertext| is invalid (but of the
+     * correct length), |out_shared_secret| is filled with a key that will always be
+     * the same for the same |ciphertext| and |private_key|, but which appears to be
+     * random unless one has access to |private_key|. These alternatives occur in
+     * constant time. Any subsequent symmetric encryption using |out_shared_secret|
+     * must use an authenticated encryption scheme in order to discover the
+     * decapsulation failure.
+     * it is assumed out_shared_secret has been allocated MLKEM_SHARED_SECRET_BYTES bytes
+     */
+    int ossl_mlkem768_decap(uint8_t *out_shared_secret,
+                            const uint8_t *ciphertext, size_t ciphertext_len,
+                            const ossl_mlkem768_private_key *private_key);
+
+    /*
+     * ossl_mlkem768_recreate_public_key recreates a fully formed ossl_mlkem768_public_key
+     * from an input |encoded_public_key| of size ossl_mlkem768_PUBLIC_KEY_BYTES.
+     * |pub| is expected to point to an allocated memory area of
+     * sizeof(ossl_mlkem768_public_key)
+     */
+    int ossl_mlkem768_recreate_public_key(const uint8_t *encoded_public_key,
```

Comment:
This is the same as BoringSSL's `ossl_mlkem768_parse_public_key`, and so I think we should match the BoringSSL name to stay consistent. I will be adding ML-KEM 768 parse/marshal private/public keys with https://github.com/openssl/private/issues/704

Reply:
OK for me, @andrewkdinh -- please note (other) API changes pending as per https://github.com/baentsch/openssl/pull/2 (review/feedback welcome!), though.